rmi and build cache miss performance improvements

runcom opened the pull request:
This patch also adds integration tests for images and rmi commands that weren't covered anywhere AFIK

Fixes #16771.

runcom committed:
integration-cli: docker_cli_build_test: check error before defer

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

runcom committed:
graph: add parent img refcount for faster rmi

also fix a typo in pkg/truncindex package comment

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

runcom committed:
daemon: faster image cache miss detection

Lookup the graph parent reference to detect a builder cache miss before
looping the whole graph image index to build a parent-children tree.

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

runcom commented:
ping @icecrime @LK4D4

LK4D4 commented:
@runcom need rebase btw :)

runcom commented:
@LK4D4 rebased

coolljt0725 commented on graph/graph.go:
```diff
@@ -424,11 +462,10 @@ func (graph *Graph) Release(sessionID string, layerIDs ...string) {
 	heads := make(map[string]*image.Image)
-	byParent := graph.ByParent()
 	graph.walkAll(func(image *image.Image) {
 		// If it's not in the byParent lookup table, then
```
Maybe `If it's not in the byParent lookup table` this comment should be removed? since there is 
no `byParent` lookup table  :)

runcom replied:
yup thanks @coolljt0725 :)

runcom committed:
graph: ensure _tmp dir is always removed

Also remove unused func `newTempFile` and prevent a possible deadlock
between pull_v2 `attemptIDReuse` and graph `register`

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

runcom commented:
I've also added https://github.com/runcom/docker/commit/84ca81ca4b0ff399fd5bf545043f5c6ba56632b8 which is fixing `graph/_tmp` leak

aaronlehmann commented on daemon/daemon.go:
```diff
@@ -1142,6 +1142,14 @@ func (daemon *Daemon) GetRemappedUIDGID() (int, int) {
+	if !daemon.Graph().HasChildren(imgID) {
+		return nil, nil
+	}
+
```
`HasChildren` calls `ByParent` internally. `ByParent` returns a map indexed by the parent ID, which is similar to what this function constructs below. Couldn't this function call `ByParent`, check `len(parents[imgID])`, and if it's nonzero, use the map returned by `ByParent` to find the earliest created child? That would avoid an extra walk of the graph.

aaronlehmann replied:
Sorry, missed your change to `HasChildren`, as it's later in the diff. This makes more sense now. I still think using `ByParent` would make more sense than `Map` in this function, but it's unrelated to your change.

runcom replied:
@aaronlehmann btw I think you're right and we could remove this whole for:
```
imageMap := make(map[string]map[string]struct{})
 	for _, img := range images {
 		if _, exists := imageMap[img.Parent]; !exists {
 			imageMap[img.Parent] = make(map[string]struct{})
 		}
 		imageMap[img.Parent][img.ID] = struct{}{}
 	}
```
and assign the map from `ByParent` to `images` (some lines below), am I correct?

runcom replied:
so there would be only 1 loop to get the parent->children tree (in `ByParent`)

aaronlehmann replied:
Yes, that's what I had in mind. `ByParent` returns a flat list of children instead of a nested map, but it looks like it doesn't matter since `ImageGetCached` just loops over the map keys in the nested map it creates.

runcom replied:
no ok, I know had a look at the code :)

aaronlehmann commented on graph/graph.go:
```diff
@@ -119,12 +120,13 @@ func NewGraph(root string, driver graphdriver.Driver, uidMaps, gidMaps []idtools
+		retained:   &retainedLayers{layerHolders: make(map[string]map[string]struct{})},
+		uidMaps:    uidMaps,
+		gidMaps:    gidMaps,
+		parentRefs: make(map[string]int),
```
Please add a comment explaining that accesses to this map need to be protected by `graph.imageMutex.Lock` on the key.

runcom replied:
:+1:

runcom replied:
done

tiborvass replied:
@runcom I don't see the comment :S

runcom replied:
hahah

aaronlehmann commented on graph/graph.go:
```diff
@@ -292,6 +308,11 @@ func (graph *Graph) Register(img *image.Image, layerData io.Reader) (err error)
+	graph.imageMutex.Lock(img.Parent)
+	graph.parentRefs[img.Parent]++
+	graph.imageMutex.Unlock(img.Parent)
+
```
I think there is a possible deadlock here given that `attemptIDReuse` (see https://github.com/docker/docker/pull/16952/files#diff-7e2a216082b6bd557c7ecebd68305e2eR576) may lock images in a different order.

cc @tonistiigi

tonistiigi replied:
I'm not sure I get how this is a deadlock if you only lock one image at a time.

aaronlehmann replied:
`Register` holds a second lock: https://github.com/docker/docker/pull/16890/files#diff-6291e801306012e55f373a86213ee4d5L244

tonistiigi replied:
so you want to check that `img.ID != img.Parent` ?

aaronlehmann commented on integration-cli/docker_cli_images_test.go:
```diff
@@ -197,3 +196,42 @@ func (s *DockerSuite) TestImagesEnsureDanglingImageOnlyListedOnce(c *check.C) {
+	c.Assert(err, check.IsNil)
+
+	split := strings.Split(out, "\n")
+	intermediate := strings.TrimSpace(split[5][7:])
```
Why the numbers 5 and 7? This probably deserves a comment (or parsing that's less sensitive to the output format).

runcom replied:
updated with comment
need intermediate img id, i don't know any other way to get one here

aaronlehmann commented on integration-cli/docker_cli_images_test.go:
```diff
@@ -197,3 +196,42 @@ func (s *DockerSuite) TestImagesEnsureDanglingImageOnlyListedOnce(c *check.C) {
+	if strings.Contains(out, intermediate) {
+		c.Fatalf("images shouldn't show non-heads images, got %s in %s", intermediate, out)
+	}
+	if !strings.Contains(out, head[:12]) {
```
Use `stringid.shortLen` instead of 12.

runcom replied:
it's not exported and lots of code deal with `:12`, probably will need a whole refactor if we decide to export shortlen from stringid

runcom replied:
I'll make a follow PR to fix this in the whole project (this has enough commits)

LK4D4 replied:
I agree, this PR already contains unrelated changes.

aaronlehmann commented on integration-cli/docker_cli_images_test.go:
```diff
@@ -197,3 +196,42 @@ func (s *DockerSuite) TestImagesEnsureDanglingImageOnlyListedOnce(c *check.C) {
+	c.Assert(err, check.IsNil)
+
+	out, _ := dockerCmd(c, "images")
+	if !strings.Contains(out, id[:12]) {
```
Same here - use `stringid.shortLen`.

tonistiigi commented on graph/graph.go:
```diff
@@ -174,6 +179,20 @@ func (graph *Graph) restore() error {
 	for _, v := range dir {
 		id := v.Name()
 		if graph.driver.Exists(id) {
+			pth := filepath.Join(graph.root, id, "json")
```
Why not `graph.loadImage(id)`? After #16953 the parent reference may not be in the raw json file. "json" is a constant `jsonFileName`.

runcom replied:
fair, i'll change it

tiborvass commented on graph/graph.go:
```diff
@@ -174,6 +179,20 @@ func (graph *Graph) restore() error {
+			}
+			defer jsonSource.Close()
+			decoder := json.NewDecoder(jsonSource)
+			img := &image.Image{}
```
you could leave `img` on the stack by removing `&` and doing `Decode(&img)`

aaronlehmann commented:
Now that https://github.com/docker/docker/pull/16953 has been merged, this needs to have some kind of locking to prevent `register` and `attemptIDReuse` from running at the same time. They both lock multiple images at once, but potentially obtain the locks in different orders, so there is a potential deadlock.

Aside from that, this looks good to me.

runcom commented:
@aaronlehmann can we lock the whole graph struct since `V2Puller` holds a reference to it when locking? maybe use this additional lock just for this synchronization. Any other ideas?

aaronlehmann commented:
Currently, there is a lock called `idReuseLock`:

https://github.com/docker/docker/blob/master/graph/pull_v2.go#L552

This could be renamed to something generic and moved to the `Graph` struct, to make a lock used by anything that locks more than one image mutex at a time.

But I'm wondering if it's really necessary for `register` to lock two mutexes at once. Can't it call `graph.imageMutex.Unlock(imgID)` before calling `graph.imageMutex.Lock(parent)`? That would work around the issue.

runcom commented:
> But I'm wondering if it's really necessary for register to lock two mutexes at once. Can't it call graph.imageMutex.Unlock(imgID) before calling graph.imageMutex.Lock(parent)? That would work around the issue.

This is much better, and there shouldn't be any issue in doing that

runcom commented:
also, it's not `register` that locks, but the exported `Register`, I'll see what I can do

thaJeztah commented:
ping @aaronlehmann I see this PR was updated, PTAL

aaronlehmann commented:
@thaJeztah: My last comment hasn't been addressed yet, as far as I can tell.

thaJeztah commented:
@aaronlehmann np, I thought you might've missed the update (didn't look in the code changes)

runcom commented:
@aaronlehmann went with avoiding locking twice in `register` instead of adding a lock to the graph. PTAL

aaronlehmann commented:
@runcom Graph unit tests appear to be hanging.

tonistiigi commented:
@runcom Note that some code paths like v2 pulls call `graph.register()` directly with the pull code itself holding the lock. Currently with this patch all v2 pulls deadlock because of this.

runcom commented:
Right, I'll fix that

runcom commented:
rebased and updated

cpuguy83 commented:
vet: integration-cli/docker_cli_rmi_test.go: integration-cli/docker_cli_rmi_test.go:235:1: expected statement, found '<<' (and 2 more errors)

runcom commented:
ouch rebase, fixed

runcom commented:
`TestConcurrentPullMultipleTags` seems to hang in experimental, will look into this

cpuguy83 commented:
So this worked well for me, tried pulling while doing rmi didn't see any deadlocks.
That said, I'm not confident enough to give an LGTM in this part of the codebase.

tonistiigi commented:
LGTM

runcom commented:
finally green :) ping @tiborvass @aaronlehmann

tiborvass commented:
@runcom what did you change in the last push?

runcom commented:
@tiborvass just rebased

tiborvass commented:
I tested it, it's much faster. weak lgtm for the code

cpuguy83 commented:
ping @aaronlehmann

tiborvass commented:
YOLO

tiborvass commented:
Removing impact/changelog because it seems this performance issue was introduced after 1.8. @runcom let me know if i'm wrong.

tiborvass commented:
Apparently i'm wrong :) adding it back
